src/System.Private.Xml/src/System/Xml/Serialization/ReflectionXmlSerializationReader.cs

```diff
@@ -614,10 +616,100 @@ private static void AddObjectsIntoTargetCollection(object targetCollection, List
-            SetMemberValue(o, value, memberInfo);
+            SetMemberValueDelegate result;
+            var typeMemberNameTuple = Tuple.Create(o.GetType(), memberName);
+            if (!s_SetMemberValueDelegateCache.TryGetValue(typeMemberNameTuple, out result))
+            {
+                MemberInfo memberInfo = ReflectionXmlSerializationHelper.GetMember(o.GetType(), memberName);
+                Type memberType;
+                if (memberInfo is PropertyInfo propInfo)
+                {
+                    memberType = propInfo.PropertyType;
+                }
+                else if (memberInfo is FieldInfo fieldInfo)
+                {
+                    memberType = fieldInfo.FieldType;
+                }
+                else
+                {
+                    throw new InvalidOperationException(SR.Format(SR.XmlInternalError));
+                }
+
+                var typeMemberTypeTuple = Tuple.Create(o.GetType(), memberType);
+                Func<MemberInfo, SetMemberValueDelegate> getSetMemberValueDelegateWithType;
+                if (!s_GetSetMemberValueDelegateWithTypeCache.TryGetValue(typeMemberTypeTuple, out getSetMemberValueDelegateWithType))
+                {
+                    MethodInfo getSetMemberValueDelegateWithTypeGenericMi = typeof(ReflectionXmlSerializationReader).GetMethod(nameof(GetSetMemberValueDelegateWithType), BindingFlags.Static | BindingFlags.NonPublic);
+                    MethodInfo getSetMemberValueDelegateWithTypeMi = getSetMemberValueDelegateWithTypeGenericMi.MakeGenericMethod(o.GetType(), memberType);
+                    getSetMemberValueDelegateWithType = (Func<MemberInfo, SetMemberValueDelegate>)getSetMemberValueDelegateWithTypeMi.CreateDelegate(typeof(Func<MemberInfo, SetMemberValueDelegate>));
+                    s_GetSetMemberValueDelegateWithTypeCache.TryAdd(typeMemberTypeTuple, getSetMemberValueDelegateWithType);
+                }
+
+                result = getSetMemberValueDelegateWithType(memberInfo);
+                s_SetMemberValueDelegateCache.TryAdd(typeMemberNameTuple, result);
+            }
+
+            return result;
+        }
+
+        private static SetMemberValueDelegate GetSetMemberValueDelegateWithType<TObj, TParam>(MemberInfo memberInfo)
+        {
+            if (typeof(TObj).IsValueType)
+            {
+                if (memberInfo is PropertyInfo propInfo)
+                {
+                    return delegate (object o, object p)
+                    {
+                        propInfo.SetValue(o, p);
+                    };
+                }
+                else if (memberInfo is FieldInfo fieldInfo)
+                {
+                    return delegate (object o, object p)
+                    {
+                        fieldInfo.SetValue(o, p);
+                    };
+                }
+
+                throw new InvalidOperationException(SR.Format(SR.XmlInternalError));
```

Comment:
This is the same exception which would would have been thrown originally when calling SetMemberValue so the behavior hasn't changed here.

---
In reply to: [122582119](https://github.com/dotnet/corefx/pull/21155#discussion_r122582119) [](ancestors = 122582119)